ExpressMapper NET40/MappingServiceBase.cs

```diff
@@ -423,28 +412,56 @@ protected LoopExpression CreateLoopExpression(Expression doMoveNextSrc, BlockExp
             {
                 return Expression.Call(destColl, destList.GetMethod("ToArray"));
             }
-            if (!destPropType.IsGenericType) return destColl;
+
+            if (!destPropType.IsGenericType && GetEnumerableInterface(destPropType) == null)
+            {
+                return destColl;
+            }
 
             if (typeof(IQueryable).IsAssignableFrom(destPropType))
             {
                 return Expression.Call(typeof(Queryable), "AsQueryable", new[] { destType }, destColl);
             }
 
-            if (destPropType.IsInterface && destColl.Type.IsSubclassOf(destPropType))
-            {
+            if (destPropType.IsInterface && destPropType.IsAssignableFrom(destColl.Type)) {
+                // This will handle any destination interface implemented by List<T>
                 return destColl;
             }
 
-            if (destPropType.IsClass)
-            {
+            ConstructorInfo ctor = null;
 
+            if (destPropType.IsInterface)
+            {
+                // We are targeting an interface type, we need to find a compatible collection type
+                // We could look for a loaded type that implements the target interface and has an appropriate
+                // constructor, but that is a bit too much magic for now.
+                throw new NotImplementedException(
+                    "Destination interface type " + destPropType.FullName + " is not supported yet" );
             }
+            else
+            {
+                ctor = destPropType.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault(
+                    ci => {
+                        var param = ci.GetParameters();
+                        return param.Length == 1 && param[0].ParameterType.IsAssignableFrom(destList);
+                    } );
 
-            return destPropType.IsClass ? Expression.New(destPropType.GetConstructor(new Type[] { destList }), destColl) : destColl;
+                if (ctor == null)
+                {
+                    throw new Exception(
+                        "Could not find a constructor on " + destPropType.Name + " that accepts " + destList);

```

Comment:
I am not sure i follow what you mean? What if we have the following:

public class MyClass : IEnumerable<T> {
  public MyClass ( int count ){}
}

It is a collection, but it does not have the appropriate ctor. I throw the exception here to make it easier to understand why it failed.